lib/pwnlib/tubes/tube.rb

```diff
@@ -101,13 +104,17 @@ def recvpred(timeout: nil)
 
               c = recv(1)
 
-              return '' if c.empty?
+              next if c.empty?
               data << c
             end
             data.slice!(0..-1)
-          ensure
```

Comment:
Why this?
Original code seems correct

Reply:
This might be a disagreement on what the function is supposed to do. In my opinion, it should not return until the timeout expires unless it has a match with the predicate. Just because a particular recv(1) didn't return anything doesn't mean that one .5s later won't.

Swap from rescue->raise is different. That's because we must not unrecv(data) if we match the predicate. With ensure, we always unrecv.

Reply:
But recv_raw is a block call 
The situation you mentioned should never happen